### PR TITLE
Google OAuth (towards #121)

### PR DESCRIPTION
**Pull request by devmrfitz**
# Description

Added Google OAuth

## Related issues
#121 
## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).

# Checklist

- [ ] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [ ] The pull request is for the branch `develop`
- [ ] A new analyzer or connector was added, in which case:
    - [ ] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](./Advanced-Usage.md) was updated (in case the analyzer/connector provides additional optional configuration).
    - [ ] Secrets were added in [env_file_app_template](https://github.com/intelowlproject/IntelOwl/blob/master/docker/env_file_app_template), [env_file_app_ci](https://github.com/certego/IntelOwl/blob/master/docker/env_file_app_ci) and in the [Installation](./Installation.md) docs, if necessary.
    - [ ] If the analyzer/connector requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [ ] If a File analyzer was added, its name was explicitly defined in [test_file_scripts.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_file_scripts.py) (not required for Observable Analyzers).
- [x] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [ ] The tests gave 0 errors.
- [ ] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] The commits were squashed into a single one (optional, they will be squashed anyway by the maintainer)
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check `CONTRIBUTE.md`).
  
### Important Rules
- If your changes decrease the overall tests coverage (you will know after the Codecov CI job is done), you should add the required tests to fix the problem
- Everytime you make changes to the PR and you think the work is done, you should explicitly ask for a review

# Real World Example

Please delete if the PR is for bug fixing.
Otherwise, please provide the resulting raw JSON of a finished analysis (and, if you like, a screenshot of the results). This is to allow the maintainers to understand how the analyzer works.

### CODE REVIEW

**Comment by devmrfitz** [comment]
Should I add the Google logo in "Google Sign In" button?

File: intel_owl/settings/security.py
Comment (0ssigeno) [inline intel_owl/settings/security.py:18]: @eshaan7 can we actually remove this without having any issue?

File: intel_owl/settings/auth.py
Comment (0ssigeno) [inline intel_owl/settings/auth.py:18]: Can we move the radius/ldap in this file too? So we have all the auth together

**Review by 0ssigeno — COMMENTED** [review]
Why didn't you use [django-allauth](https://github.com/pennersr/django-allauth)

File: intel_owl/settings/security.py
Comment (devmrfitz) [reply intel_owl/settings/security.py:18]: I have sent a message on Slack regarding this as well

**Comment by devmrfitz** [comment]
> Why didn't you use [django-allauth](https://github.com/pennersr/django-allauth)?

`allauth` (along with `dj-rest-auth`) is an extensive authentication solution in itself. However, we are already using `django-rest-durin` to handle the main authentication flow, and hence bringing the former would be an overkill and also require a lot of modifications to the core workings of `dj-rest-auth`.

File: intel_owl/settings/security.py
Comment (0ssigeno) [reply intel_owl/settings/security.py:18]: Just adding another pointer :D

**Comment by 0ssigeno** [comment]
I'm reading a little bit about `allauth` to see which changes would that requires on our side. Your explanation makes sense, and I do not see any particular issue on following your implementation, just wanted to double check

File: intel_owl/settings/auth.py
Comment (devmrfitz) [reply intel_owl/settings/auth.py]: Done

**Comment by devmrfitz** [comment]
> I'm reading a little bit about `allauth` to see which changes would that requires on our side. Your explanation makes sense, and I do not see any particular issue on following your implementation, just wanted to double check

The interesting thing is that I initially did a partial implementation using `dj-rest-auth` and `allauth` before going to the current implementation. xD

**Review by 0ssigeno — COMMENTED** [review]
Can we have some tests on the views/urls ?

**Comment by 0ssigeno** [comment]
I'm ok with this choice. If we will ever need to support _many different_ external auth, we can go on the `allauth` road

File: intel_owl/settings/auth.py
Comment (devmrfitz) [reply intel_owl/settings/auth.py:50]: Makes sense. What would be better, something to enable/disable all alternative authentications or individual controls (one for let's say Google OAuth, another for Github OAuth, etc)?

**Comment by devmrfitz** [comment]
> Can we have some tests on the views/urls ?

That's something I'm having problems working out. How exactly do we automate the testing of Google (and possibly other) OAuths and is it not a violation of their policies?

**Comment by 0ssigeno** [comment]
For the urls, at least that a 200 code is returned.
For the `GoogleLoginCallbackView`, you can mock `oauth.google.authorize_access_token` and check the that the result matches that expectation. You can test the class itself instead of the `login` flow, at least we will have some kind of coverage

**Comment by 0ssigeno** [comment]
If you can test entire flow, mocking here and there, i will be very glad. But, as you said, do not query directly google for that, let's try to mock their answer

**Comment by 0ssigeno** [comment]
Dw about the `do not merge thing`; if a PR is in draft mode, we won't merge it. Once you change that, once all the comments have resolved and everyone had time to make their review, we will merge it. Just FYI for the future :D

**Comment by devmrfitz** [comment]
I'm sorry for the weird PR title and message. The reason for that is the fact that when I use `DANGEROUSLY_DISABLE_HOST_CHECK=true npm start`, the backend is on port 80 while frontend is on port 3001, and the backend does not know this, hence the redirects are not working properly. I'll replace this with the proper relative URLs once the PR is ready to be merged.

**Comment by devmrfitz** [comment]
> Dw about the `do not merge thing`; if a PR is in draft mode, we won't merge it. Once you change that, once all the comments have resolved and everyone had time to make their review, we will merge it. Just FYI for the future :D

Yeah. However, I was afraid that I might mark it `Ready for Review` without remembering this change, as it wouldn't pop up in normal testing. Hence I did something unusual so it hits me when I'm undoing that

File: intel_owl/settings/auth.py
Comment (devmrfitz) [reply intel_owl/settings/auth.py]: @0ssigeno

**Comment by devmrfitz** [comment]
@0ssigeno I'm done with the tests. Any other changes you propose?

File: intel_owl/settings/auth.py
Comment (drosetti) [reply intel_owl/settings/auth.py]: I think individual controls is a better solution, it will require more env vars, but it will be more accurate

**Comment by drosetti** [comment]
@devmrfitz

@0ssigeno  is out of office and he will come back next week, I'm reading your messages in this PR and I'm going to review you code, if you have questions or need some support feel free to tag me :D

**Review by drosetti — COMMENTED** [review]
Good work! I have some questions/requests:

* Please add the `close` tag to close the related issue
* Currently the Google auth works in case the user already exist, if user is missing it raise an error. I think we could create the user in case it doesn't exist. I'm not but I think that also ldap authentication works in this way.
* I think we could move the `authentication` package outside `api_app`

What do you think ?

**Comment by devmrfitz** [comment]
> * Currently the Google auth works in case the user already exist, if user is missing it raise an error. I think we could create the user in case it doesn't exist. I'm not but I think that also ldap authentication works in this way.

IntelOwl's current access model is that only pre-authorized people can login. Even in LDAP's case, only pre-authorized user's have their accounts in the LDAP server.
However, if we allow account creation through Google OAuth, anyone with a Gmail ID would get access to it.

File: tests/auth/__init__.py
Comment (devmrfitz) [reply tests/auth/__init__.py:26]: Done

**Comment by devmrfitz** [comment]
> * Please add the `close` tag to close the related issue

That issue is for multiple external authentications. Should I create a separate one for Google?

**Comment by devmrfitz** [comment]
> * I think we could move the `authentication` package outside `api_app`

Done

**Comment by 0ssigeno** [comment]
Is correct that the user is not created if the google auth fails.
Giving a fast double check, but imho is mergeable

**Review by 0ssigeno — APPROVED** [review]

**Review by drosetti — APPROVED** [review]
> IntelOwl's current access model is that only pre-authorized people can login. Even in LDAP's case, only pre-authorized user's have their accounts in the LDAP server.
> However, if we allow account creation through Google OAuth, anyone with a Gmail ID would get access to it.

No you are right, you work is correct

> That issue is for multiple external authentications. Should I create a separate one for Google?

You are right again, we can leave the issue open, I will write something about the authentication  with Google

You PR is ok, I approve it and i'm going to merge it, good work!

**Comment by 0ssigeno** [comment]
@devmrfitz once the tests pass we are going to merge it

**Comment by devmrfitz** [comment]
@0ssigeno Tests done

**Pull request merged**
